set.c

```diff
@@ -0,0 +1,2149 @@
+ *
+ *  This is aliased to #===, so it is usable in +case+ expressions:
+ *
+ *    case :apple
+ *    when Set[:potato, :carrot]
+ *      "vegetable"
+ *    when Set[:apple, :banana]
+ *      "fruit"
+ *    end
+ *    # => "fruit"
+ *
+ *  See also Enumerable#include?
+ */
+static VALUE
+set_i_include(VALUE set, VALUE item)
+{
+    return RBOOL(RSET_IS_MEMBER(set, item));
+}
+
+struct set_merge_args {
+  VALUE set;
+  set_table *into;
+};
+
+static int
+set_merge_i(st_data_t key, st_data_t data)
+{
+    struct set_merge_args *args = (struct set_merge_args *)data;
+    key = set_stringify(args->into, (VALUE)key);
+    RB_OBJ_WRITTEN(args->set, Qundef, (VALUE)key);
+    set_insert(args->into, key);
+    return ST_CONTINUE;
+}
+
+static VALUE
+set_merge_block(RB_BLOCK_CALL_FUNC_ARGLIST(key, set))
+{
+    set_table *into = RSET_TABLE(set);
+    key = set_stringify(into, key);
+    RB_OBJ_WRITTEN(set, Qundef, key);
+    set_insert(into, (st_data_t)key);
+    return key;
+}
+
+static void
+set_merge_enum_into(VALUE set, VALUE arg)
+{
+    if (rb_obj_is_kind_of(arg, rb_cSet)) {
+        struct set_merge_args args = {
+            .set = set,
+            .into = RSET_TABLE(set)
+        };
+        set_iter(arg, set_merge_i, (st_data_t)&args);
+    }
+    else if (RB_TYPE_P(arg, T_ARRAY)) {
+        long len = RARRAY_LEN(arg);
+        if (RARRAY_LEN(arg) != 0) {
+            set_table *into = RSET_TABLE(set);
+            VALUE key;
+            RARRAY_PTR_USE(arg, ptr, {
+                for(; len > 0; len--, ptr++) {
+                    key = set_stringify(into, *ptr);
+                    RB_OBJ_WRITTEN(set, Qundef, key);
+                    set_insert(into, (st_data_t)key);
+                }
+            });
+        }
+    }
+    else {
+        rb_block_call(arg, enum_method_id(arg), 0, 0, set_merge_block, (VALUE)set);
+    }
+}
+
+/*
+ *  call-seq:
+ *    merge(*enums, **nil) -> self
+ *
+ *  Merges the elements of the given enumerable objects to the set and
+ *  returns self.
+ */
+static VALUE
+set_i_merge(int argc, VALUE *argv, VALUE set)
+{
+    if (rb_keyword_given_p()) {
+        rb_raise(rb_eArgError, "no keywords accepted");
+    }
+    rb_check_frozen(set);
+
+    int i;
+
+    for (i=0; i < argc; i++) {
+        set_merge_enum_into(set, argv[i]);
+    }
+
+    return set;
+}
+
+static VALUE
+set_reset_table_with_type(VALUE set, const struct st_hash_type *type)
+{
+    rb_check_frozen(set);
+
+    struct set_object *sobj;
+    TypedData_Get_Struct(set, struct set_object, &set_data_type, sobj);
+    set_table *old = &sobj->table;
+
+    size_t size = set_table_size(old);
+    if (size > 0) {
+        set_table *new = set_init_table_with_size(NULL, type, size);
+        struct set_merge_args args = {
+            .set = set,
+            .into = new
+        };
+        set_iter(set, set_merge_i, (st_data_t)&args);
+        set_free_embedded(sobj);
+        memcpy(&sobj->table, new, sizeof(*new));
```

Comment:
We might have created new references here, no? If so we need to trigger write barriers.

Reply:
Not needed in this function as the write barrier are added in `set_merge_i` (called a couple lines above).